lib/MeshBVHNode.js

```diff
@@ -16,53 +16,76 @@ class MeshBVHNode {
 		this.geometry = null;
 		this.tris = null;
 		this.children = null;
+		this.splitAxis = - 1;
 
 	}
 
-	intersectsRay( ray ) {
-
-		arrayToSphere( this.boundingData, boundingSphere );
-
-		if ( ! ray.intersectsSphere( boundingSphere ) ) return false;
+	intersectRay( ray, target ) {
 
 		arrayToBox( this.boundingData, boundingBox );
 
-		return ray.intersectsBox( boundingBox );
+		return ray.intersectBox( boundingBox, target );
 
 	}
 
 	raycast( mesh, raycaster, ray, intersects, seenFaces ) {
 
-		if ( ! this.intersectsRay( ray ) ) return;
-
 		if ( this.tris ) intersectTris( mesh, this.geometry, raycaster, ray, this.tris, intersects, seenFaces );
-		else this.children.forEach( c => c.raycast( mesh, raycaster, ray, intersects, seenFaces ) );
+		else this.children.forEach( c => {
+
+			if ( c.intersectRay( ray, boxIntersection ) )
+				c.raycast( mesh, raycaster, ray, intersects, seenFaces );
+
+		} );
 
 	}
 
 	raycastFirst( mesh, raycaster, ray ) {
 
-		if ( ! this.intersectsRay( ray ) ) return null;
-
 		if ( this.tris ) {
 
 			return intersectClosestTri( mesh, this.geometry, raycaster, ray, this.tris );
 
 		} else {
 
-			const c1 = this.children[ 0 ];
-			arrayToSphere( c1.boundingData, boundingSphere );
-			intersectvec.subVectors( boundingSphere.center, ray.origin );
-			const c1dist = intersectvec.length() * intersectvec.dot( ray.direction );
+			// consider the position of the split plane with respect to the oncoming ray; whichever direction
+			// the ray is coming from, look for an intersection among that side of the tree first
+
+			const leftToRight = ray.direction[ xyzFields[ this.splitAxis ] ] <= 0;
+			const c1 = leftToRight ? this.children[ 0 ] : this.children[ 1 ];
+			const c2 = leftToRight ? this.children[ 1 ] : this.children[ 0 ];
+
+			const c1Intersection = c1.intersectRay( ray, boxIntersection );
+			const c1Result = c1Intersection ? c1.raycastFirst( mesh, raycaster, ray ) : null;
+			const c2Intersection = c2.intersectRay( ray, boxIntersection );
```

Comment:
The raycast already seems really performant so it's not a big deal but I believe this check can be faster if you're interested.

This `intersectRay` function for the box will check all six faces of the box when really you only need to check a _single_ component of _one_ of the planes of the box. So if the split axis is `z` then you should only have to check if the `z` component of the split plane is greater than or less than the `z` component of the intersection (depending on the direction of the ray). That should also let you get rid of the call to `distanceToSquared` below.

Reply:
Hmm. I see what you're getting at, I'll think about it.

Reply:
I'll plan to merge this as it is for now and we can look at this in another PR if it's interesting.